docs/03-code-internals/19-service-objects.md

```diff
@@ -0,0 +1,602 @@
+
+You can also provide an `ActiveRecord` relation. The step won’t load records but will determine if the relation will return any records. If there are none, the execution flow will stop.
+
+Sometimes, you need to fetch a model (or a collection of models), but it’s ok if it’s empty. For those cases, you can use the `optional: true` option allowing the execution flow to continue even if the model returns a falsy value.
+
+### `policy`
+
+This step will execute the method of the same name. You can put arbitrary code here, and the execution flow will stop if the return value is falsy.
+
+Usually, a policy is related to some state on one of the service models and/or to the current user (if any).
+
+### `contract`
+
+This is one of the most powerful steps. Its main purpose is to validate the incoming data before feeding it to the models and to the service at a more global level. This is actually `ActiveModel` validations but applied to the incoming parameters.
+
+This step will run coercions and validations defined in the provided block. If the contract isn’t valid (at least one validation failed), then the execution flow will stop.
+
+### `transaction`
+
+This step is a bit special, as it will wrap any other steps defined in its block inside a SQL transaction.
+
+### Steps arguments
+
+Each step is called with the service context. To access a value in it, just provide its key as a keyword argument.
+
+## The context object
+
+The only state a service maintain is its context object. This is where each step can put data to be used by other steps. Most of the time, you don’t need to access the context directly, as specialized steps will store the proper data for you. But sometimes it’s necessary. In those cases, it’s just a matter of using the context like you would with a hash:
+```ruby
+def first_step
+  context[:my_special_key] = "My special value"
+end
+
+def second_step(my_special_key:)
+  # do something with `my_special_key`
+end
+```
+
+## Handling service results
+
+Once a service has been called, it will return a result object, to know whether the call was a success. In the case of a failure, the result object can be inspected to know what failed and why.
+
+Each step will store its outcome in the result object, accessible through special keys (like `result.model.user` for example). While this is nice, this would be tedious to manually check the result object. That’s why there’s a built-in feature allowing to run the service, match step outcomes and act upon results using a custom DSL.
+
+This feature makes the use of a service a breeze. Continuing with our `User::UpdateUsername` service, this is how we could use it inside a controller (but it can work anywhere, not just in controllers):
+```ruby
+def update
+  User::UpdateUsername.call(service_params) do
+    on_success { render(json: success_json) }
+    on_failure { render(json: failed_json, status: 422) }
```

Comment:
Is there a way to get the error messages on any failure here? Similar to the line below? Or do we need to set up one `on_failed_x` for each thing?

Reply:
You have access to `result` so you can inspect the result object manually, but it’s not very practical. The idea is to use `on_failed_*` to match any specific error you can have, since you know about them. `on_failure` is a generic matcher, to use either to catch a failure not handled by the other matchers or if you don’t really care about a specific failure and just want to always act the same way when there’s a failure.

Tell me if that wasn’t clear enough :sweat_smile: